src/main/java/io/airlift/compress/zstd/ZstdFrameDecompressor.java

```diff
@@ -208,6 +217,342 @@ public int decompress(
+        final int blockHeader;
+
+        final int inputConsumed;
+        final int outputBufferUsed;
+
+        final int inputRequired;
+        final int requestedOutputSize;
+
+        private DecompressState()
+        {
+            this.state = State.READ_FRAME_MAGIC;
+            this.atStoppingPoint = false;
+            this.frameHeader = null;
+            this.blockHeader = -1;
+            this.inputConsumed = 0;
+            this.outputBufferUsed = 0;
+            this.inputRequired = 0;
+            this.requestedOutputSize = 0;
+        }
+
+        DecompressState(State state, FrameHeader frameHeader, int blockHeader, int inputConsumed, int outputBufferUsed, int inputRequired, int requestedOutputSize)
+        {
+            this.state = requireNonNull(state, "state is null");
+            this.atStoppingPoint = state == State.READ_FRAME_MAGIC;
+            this.frameHeader = frameHeader;
+            this.blockHeader = blockHeader;
+
+            checkArgument(inputConsumed >= 0, "inputConsumed is negative");
+            this.inputConsumed = inputConsumed;
+            checkArgument(outputBufferUsed >= 0, "outputBufferUsed is negative");
+            this.outputBufferUsed = outputBufferUsed;
+            checkArgument(inputRequired >= 0, "inputRequired is negative");
+            this.inputRequired = inputRequired;
+            checkArgument(requestedOutputSize >= 0, "requestedOutputSize is negative");
+            this.requestedOutputSize = requestedOutputSize;
+        }
+
+        public boolean isAtStoppingPoint()
+        {
+            return atStoppingPoint;
+        }
+
+        public int getInputConsumed()
+        {
+            return inputConsumed;
+        }
+
+        public int getOutputBufferUsed()
+        {
+            return outputBufferUsed;
+        }
+
+        public int getInputRequired()
+        {
+            return inputRequired;
+        }
+
+        public int getRequestedOutputSize()
+        {
+            return requestedOutputSize;
+        }
+
+        @Override
+        public String toString()
+        {
+            return new StringJoiner(", ", DecompressState.class.getSimpleName() + "[", "]")
+                    .add("state=" + state)
+                    .add("frameHeader=" + frameHeader)
+                    .add("blockHeader=" + blockHeader)
+                    .add("inputConsumed=" + inputConsumed)
+                    .add("outputBufferUsed=" + outputBufferUsed)
+                    .add("inputRequired=" + inputRequired)
+                    .add("requestedOutputSize=" + requestedOutputSize)
+                    .toString();
+        }
+    }
+
+    // current window buffer
+    private Object windowBase = new byte[0];
+    private long windowAddress = ARRAY_BYTE_BASE_OFFSET;
+    private long windowLimit = ARRAY_BYTE_BASE_OFFSET;
+    private long windowPosition = ARRAY_BYTE_BASE_OFFSET;
+
+    private XxHash64 partialHash;
+
+    public DecompressState partialDecompress(
```

Comment:
Does this implementation share anything with the non-partial decompressor? If not, I'd move it into a separate class (e.g., ZstdIncrementalFrameDecompressor or ZstdResumableFrameDecompressor